### PR TITLE
[dataflow] Add backward analysis in data-flow framework

### PR DESCRIPTION
**Pull request by CharlesZ-Chen**
Major modifications:
- extract original `Analysis`  as a general interface
- add `AbstractAnalysis` as common code base of `BackwardAnalysis` and `ForwardAnalysis`
- implement `BackwardAnalysis`
- extract `TransferFunction` as a general interface, add `BackwardTransferFunction` and `ForwardTransferFunction`
- move `env` and `types` From original `Analysis` to `CFAbstractAnalysis` as they are checker-framework specific fields
- small changes in checker-framework to adapt changes in dataflow-framework

### CODE REVIEW

File: dataflow/src/org/checkerframework/dataflow/analysis/AbstractAnalysis.java [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+                V extends AbstractValue<V>, S extends Store<S>, T extends TransferFunction<V, S>>
+        implements Analysis<V, S, T> {
+
+    /** Is the analysis currently running? */
+    protected boolean isRunning = false;
+
+    /** The control flow graph to perform the analysis on. */
+    protected ControlFlowGraph cfg;
+
+    protected final Direction direction;
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/analysis/AbstractAnalysis.java]: Like for the other fields, add some documentation.

File: dataflow/src/org/checkerframework/dataflow/analysis/AbstractAnalysis.java [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+    /**
+     * The current transfer input when the analysis is running.
+     */
+    protected TransferInput<V, S> currentInput;
+
+    public AbstractAnalysis(Direction direction) {
+        this.direction = direction;
+    }
+
+    protected abstract void initInitialInputs();
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/analysis/AbstractAnalysis.java]: Add some documentation for what the method does and when it is called.

File: dataflow/src/org/checkerframework/dataflow/analysis/AbstractAnalysis.java [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+     * Runs the analysis again within the block of {@code node} and returns the
+     * store at the location of {@code node}. If {@code before} is true, then
+     * the store immediately before the {@link Node} {@code node} is returned.
+     * Otherwise, the store after {@code node} is returned.
+     */
+    protected abstract S runAnalysisFor(
+            Node node, boolean before, TransferInput<V, S> transferInput);
+
+    @Override
+    /** Is the analysis currently running? */
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/analysis/AbstractAnalysis.java]: I would put the Javadoc before the `@Override`.

File: dataflow/src/org/checkerframework/dataflow/analysis/Analysis.java [hunk 4]
```diff
@@ -1,752 +1,61 @@
...
-
-                        // propagate store to successors
-                        Block succ = rb.getSuccessor();
-                        assert succ != null
-                                : "regular basic block without non-exceptional successor unexpected";
-                        propagateStoresTo(
-                                succ, lastNode, currentInput, rb.getFlowRule(), addToWorklistAgain);
-                        break;
-                    }
+    public Direction getDirection();
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/analysis/Analysis.java]: Now that this is an interface, having good documentation for the interfaces is even more important. Most of these methods already existed, so copy over the documentation for them.

Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/analysis/Analysis.java]: Methods in interfaces are automatically `public`. You can remove it.

File: dataflow/src/org/checkerframework/dataflow/analysis/AnalysisResult.java [hunk 5]
```diff
@@ -133,6 +134,85 @@ public S getStoreBefore(Node node) {
...
+                    }
+                    if (firstNode == null) {
+                        // this block doesn't contains any node, return store in transfer input
+                        return transferInput.getRegularStore();
+                    }
+                    return analysis.runAnalysisFor(firstNode, true, transferInput);
+                }
+            default:
+                {
+                    assert false;
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/analysis/AnalysisResult.java]: Use the ErrorReporter to always raise an error, not just when assertions are enabled. Provide a message.

File: dataflow/src/org/checkerframework/dataflow/analysis/AnalysisResult.java [hunk 6]
```diff
@@ -170,69 +250,9 @@ protected S runAnalysisFor(Node node, boolean before) {
...
-                    assert false;
-                    break;
-            }
-
-            return null;
-        } finally {
-            analysis.currentNode = oldCurrentNode;
-            analysis.isRunning = false;
+        if (transferInput.analysis == null) {
+            throw new RuntimeException("transferInput contains null analysis!");
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/analysis/AnalysisResult.java]: Can you use the `ErrorReporter` instead?

File: dataflow/src/org/checkerframework/dataflow/analysis/BackwardAnalysis.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package org.checkerframework.dataflow.analysis;
+
+public interface BackwardAnalysis<
+                V extends AbstractValue<V>,
+                S extends Store<S>,
+                T extends BackwardTransferFunction<V, S>>
+        extends Analysis<V, S, T> {
+
+    public S getEntrySotre();
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/analysis/BackwardAnalysis.java]: "Sotre"? Should this be "Store"?
No visibility modifiers in interfaces.

File: dataflow/src/org/checkerframework/dataflow/analysis/BackwardTransferFunction.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package org.checkerframework.dataflow.analysis;
+
+import java.util.List;
+import org.checkerframework.dataflow.cfg.UnderlyingAST;
+import org.checkerframework.dataflow.cfg.node.ReturnNode;
+
+public interface BackwardTransferFunction<V extends AbstractValue<V>, S extends Store<S>>
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/analysis/BackwardTransferFunction.java]: Interfaces should have documentation. See what parts of the existing ForwardTF can me used for TF and BackwardTF. Put the common stuff into TF and then the specifics into the subclass javadoc.

File: dataflow/src/org/checkerframework/dataflow/cfg/CFGProcessor.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+import com.sun.source.util.TreePathScanner;
+import javax.annotation.processing.SupportedAnnotationTypes;
+import javax.lang.model.SourceVersion;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import org.checkerframework.javacutil.BasicTypeProcessor;
+import org.checkerframework.javacutil.TreeUtils;
+
+@SupportedAnnotationTypes("*")
+public class CFGProcessor extends BasicTypeProcessor {
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/CFGProcessor.java]: What is this class for?

File: dataflow/src/org/checkerframework/dataflow/cfg/CFGVisualizer.java [hunk 10]
```diff
@@ -136,8 +136,15 @@
     void visualizeStoreKeyVal(String keyName, Object value);
 
     /**
-     * Called by {@code CFAbstractStore} to visualize
-     * any information after the invocation of {@code CFAbstractStore#internalVisualize()}.
+     * visualize specific information collected according to the specific kind of Store.
+     * Currently, <code>LiveVariableStore</code> use this method to visualize
+     * @param value the value of the specific information to be visualized
+     */
+    void visualizeSotreVal(Object value);
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/CFGVisualizer.java]: Sotre -> Store?

File: dataflow/src/org/checkerframework/dataflow/cfg/CFGProcessor.java (hunk 9 above)
Comment (CharlesZ-Chen) [inline dataflow/src/org/checkerframework/dataflow/cfg/CFGProcessor.java]: This class is a refactor result on original `JavaSource2DOTCFG`. It's a bit un-related to this backward analysis refactor. I will separate it and make another separate PR.

**Comment by CharlesZ-Chen** [comment]
Hi Prof. Dietl,

I've updated and further clean up this PR as your comments. Major changes:

- add documentation in interface `Analysis`, `TransferFunction`, 'BackwardTF` and `ForwardTF`.

- refactor interface `Analysis` a bit, moving some methods down to `AbstractAnalysis`.

- clean up the usage of assertion, and replace with `ErrorReporter.errorAbort()`

- separate unrelated changes out from this PR, include `CFGVisualizer` changes and unrelated formatting changes.

I think this PR is ready for another round of code review.

Thanks,

Charles

**Comment by wmdietl** [comment]
@xingweitian Can you please take these changes and start a fresh PR?
The file structure of the CF changed significantly since then, but hopefully it is not too hard to update.
Similarly for #17 .

**Comment by xingweitian** [comment]
> @xingweitian Can you please take these changes and start a fresh PR?
> The file structure of the CF changed significantly since then, but hopefully it is not too hard to update.
> Similarly for #17 .

I think I need some time to finish the update..

File: dataflow/src/org/checkerframework/dataflow/constantpropagation/ConstantPropagationTransfer.java [hunk 11]
```diff
@@ -38,7 +38,7 @@ public ConstantPropagationStore initialStore(
     @Override
     public TransferResult<Constant, ConstantPropagationStore> visitNode(
             Node n, TransferInput<Constant, ConstantPropagationStore> p) {
-        return new RegularTransferResult<>(null, p.getRegularStore());
+        return new RegularTransferResult<>(null, p.getRegularStore().copy());
```
Comment (xingweitian) [inline dataflow/src/org/checkerframework/dataflow/constantpropagation/ConstantPropagationTransfer.java]: Why we need to call `.copy()` at here?

**Comment by xingweitian** [comment]
This is going to close since we have a new PR #99 now.

**Pull request closed without merging**